feat: Implement features from `prose-core-views` 0.4.0...0.9.0

RemiBardon opened the pull request.

nesium reviewed (commented):
Thanks, Rémi!

A few notes…

At first glimpse introducing `WebMessageHandler` seemed like a good idea to me. After thinking about this a little more deeply, I'm not so sure anymore.

I believe that introducing this new concept makes reasoning about the code harder. As an example, imagine you'd wanted to change something about the context menu without knowing  the code base too much. Without these handlers you'd _knew_ that there is no other way than that the context menu would be presented from the view - now this doesn't hold true anymore. Because it's not exactly clear what they are, these handlers are also prone to misuse. Are these views? Are they allowed to generate side-effects?

The alternative - for this feature set - would be that we'd add an extra messageReducer that receives the JS events as actions. This could be a [forEach reducer](https://pointfreeco.github.io/swift-composable-architecture/Reducer/#reducer.foreach(state:action:environment:file:line:)). The messages should be held in an `IdentifiedArray`, also because there's a lot of index searching going on already. In case encoding becomes a problem it could be done like that:

```swift
struct Obj: Encodable {
  var items: [String] = ["a", "b", "c"]

  func encode(to encoder: Encoder) throws {
    var container = encoder.unkeyedContainer()
    
    for item in self.items {
      try container.encode(item)
    }
  }
}

try print(String(data: JSONEncoder().encode(Obj()), encoding: .utf8)!)
// prints ["a","b","c"]
```

The `MessageMenu` could then be a plain struct (akin to [AlertState](https://pointfreeco.github.io/swift-composable-architecture/AlertState/)) which is not tied to a specific platform. We'd then have a custom `View` which interprets the `MessageMenu` depending on the current platform.

Example for a modified JS event handler, building on your work…

```swift
struct ToggleReactionPayload: Codable {
  var ids: [Message.ID]
  var reaction: String
}

enum MessageAction {
  case showReactions([Message.ID])
  case toggleReaction(ToggleReactionPayload)
}

struct JSEventHandler {
  var event: String
  var actionFromBody: (Any) throws -> MessageAction

  init<Payload: Decodable>(event: String, action: @escaping (Payload) -> MessageAction) {
    self.event = event
    self.actionFromBody = { body in
      guard let bodyString: String = body as? String,
            let bodyData: Data = bodyString.data(using: .utf8)
      else {
        logger.fault("JS message body should be serialized as a String")
        // TODO: Throw exception
        fatalError()
      }
      
      guard let payload = try? JSONDecoder().decode(Payload.self, from: bodyData) else {
        logger.warning("JS message body could not be decoded as `Self.Body`. Content: \(bodyString)")
        // TODO: Throw exception
        fatalError()
      }
      
      return action(payload)
    }
  }
}

final class ViewStoreScriptMessageHandler: NSObject, WKScriptMessageHandler {
  let viewStore: ViewStore<Void, MessageAction>
  let handler: JSEventHandler
  
  init(handler: JSEventHandler, viewStore: ViewStore<Void, MessageAction>) {
    self.handler = handler
    self.viewStore = viewStore
    super.init()
  }

  func userContentController(
    _ userContentController: WKUserContentController,
    didReceive message: WKScriptMessage
  ) {
    /// TODO: Handle exception
    try! self.viewStore.send(self.handler.actionFromBody(message.body))
  }
}

extension WKUserContentController {
  func addEventHandler(_ handler: JSEventHandler, viewStore: ViewStore<Void, MessageAction>) {
    let handlerName = "handler_" + handler.event.replacingOccurrences(of: ":", with: "_")
  
    self.add(
      ViewStoreScriptMessageHandler(handler: handler, viewStore: viewStore),
      name: handlerName
    )
    
    let script = """
    function \(handlerName)(content) {
      // We need to send a parameter, or the call will not be forwarded to the `WKScriptMessageHandler`
      window.webkit.messageHandlers.\(handlerName).postMessage(JSON.stringify(content));
    }
    MessagingEvent.on("\(handler.event)", \(handlerName));
    """
    
    self.addUserScript(
      WKUserScript(source: script, injectionTime: .atDocumentEnd, forMainFrameOnly: true)
    )
  }
}

// Usage
let handler = JSEventHandler(event: "message:reactions:view", action: MessageAction.showReactions)
```

What do you think? Am I missing anything?

nesium commented on Prose/ProseLib/Sources/ConversationFeature/Chat/Chat.swift:
```diff
@@ -106,19 +214,137 @@ struct Chat: NSViewRepresentable {
+    case let .copyText(id):
+      logger.trace("Copying text of \(String(describing: id))…")
+      if let message = state.messages.last(where: { $0.id == id }) {
+        NSPasteboard.general.clearContents()
```
That should be a new dependency since this is clearly a side-effect. That way we also can handle platform differences more easily.

```swift
public struct PasteboardClient {
  public var copyString: (String) -> Void

  public init(copyString: @escaping (String) -> Void) {
    self.copyString = copyString
  }
}

public extension PasteboardClient {
  static func live(pasteboard: UIPasteboard = .general) -> Self {
    .init(copyString: { pasteboard.string = $0 })
  }
}
```

RemiBardon replied:
Ok, I will do that

RemiBardon commented on Prose/ProseLib/Sources/ConversationFeature/Chat/Events/MessageMenu.swift:
```diff
@@ -0,0 +1,171 @@
+
+    var keyEquivalent: String {
+      switch self {
+      case .copyText: return "c"
```
Yes, doesn't it work for you? Are you opening the menu and pressing <kbd>C</kbd>?

RemiBardon commented on Prose/ProseLib/Sources/ProseCoreTCA/Foundation+Ext.swift:
```diff
@@ -0,0 +1,34 @@
+  /// Inserts a new element or removes it if already present.
+  /// - Parameter element: An element to insert/remove.
+  /// - Returns: `true` if the element was inserted, `false` if it was removed.
+  @discardableResult mutating func prose_toggle(_ element: Element) -> Bool {
```
> Have you thought about using a thin custom wrapper for the reactions?

No, I didn't, but I can do it, it looks better.

For the second part, I don't understand why we'd need all of this code if we're still sorting on keys (`self.items.keys.sorted()`) 🤔 Dictionary already encodes as `{"a":1,"b":2,"c":3}`, and using `self.items.keys.sorted()` doesn't solve the problem that `OrderedDictionary` solves (ordering).

Edit: Were you proposing me to create a wrapper around `OrderedDictionary` with a custom  `encode(to:)` implementation?

nesium replied:
Exactly. The encoding example was just for illustrating the use of a custom coding key.

RemiBardon replied:
Ok I have something surprising here:

```log
Encoding: 😃 [valerian@prose.org]
Encoding: 👍 [preview@prose.org]
Result: {"👍":["preview@prose.org"],"😃":["valerian@prose.org"]}
```

I encode keys in the correct order, but the `JSONEncoder` outputs them in a different order. I don't know what I'm missing.

nesium replied:
It worked in my limited test, but that may very well be the case what you're observing. Just leave `sortedKeys` enabled then.

RemiBardon replied:
I tried

```swift
import Foundation
import OrderedCollections

struct StringCodingKey: CodingKey {
  var stringValue: String

  init(stringValue: String) {
    self.stringValue = stringValue
  }
  init(_ character: Character) {
    self.init(stringValue: String(describing: character))
  }

  var intValue: Int?

  init(intValue: Int) {
    self.init(stringValue: String(describing: intValue))
    // We only want strings
//    self.intValue = intValue
  }
}

struct Obj: Encodable {
  var items: OrderedDictionary<Character, OrderedSet<String>> = .init()

  func encode(to encoder: Encoder) throws {
    var container = encoder.container(keyedBy: StringCodingKey.self)
    for (key, value) in self.items {
      try container.encode(value, forKey: StringCodingKey(key))
    }
  }
}

func test(_ obj: Obj) {
  do {
    let json = try String(data: JSONEncoder().encode(obj), encoding: .utf8)!
    print("\"\(json)\"")
  } catch {
    print(error.localizedDescription)
  }
}

test(Obj())
test(Obj(items: ["🤩": ["a", "b"]]))
test(Obj(items: ["🤩": ["a", "b"], "👍🏼": ["b", "c"]]))
test(Obj(items: ["👍🏼": ["b", "c"], "🤩": ["a", "b"]]))
```

and it gives

```log
"{}"
"{"🤩":["a","b"]}"
"{"👍🏼":["b","c"],"🤩":["a","b"]}"
"{"👍🏼":["b","c"],"🤩":["a","b"]}"
```

I think it comes from the fact that [a [JSON] object is an unordered set of name/value pairs](https://www.json.org/json-en.html). We have no warranty that the order will be preserved.

@valeriansaliou As the order is important to the logic of `prose-core-views`, could you change the API so it takes an array of objects instead? Something like:

```json
[
  {"reaction": "👍🏼", "from": ["b","c"]},
  {"reaction": "🤩", "from": ["a","b"]}
]
```

nesium replied:
I second this. Expecting dictionaries to be sorted relies on implementation details.

valeriansaliou replied:
Yes, will do today.

RemiBardon replied:
https://github.com/prose-im/prose-core-views/releases/tag/0.10.0 solves this issue. I will not use the fix in this PR, but I'll bump `prose-core-views` in a future PR.

RemiBardon commented on Prose/ProseLib/Sources/ConversationFeature/Chat/Chat.swift:
```diff
@@ -106,19 +214,137 @@ struct Chat: NSViewRepresentable {
+       let message = state.messages.last(where: { $0.id == id }),
+       message.from == loggedInUserJID
+    {
+      return .fireAndForget {
```
I'm running it on the main queue, is there another way to be sure it runs there?

RemiBardon commented:
> What do you think? Am I missing anything?

@nesium The problem with your approach is that you suppose all actions only need the message payload (serialized into the correct type). However, when the payload contains a point, we also need a reference to the `WKWebView`. I will make it so all actions take a pair of `(WKWebView, Payload)` as an argument, even if we don't need the web view sometimes. It's already what I was doing with the `func handle(_ message: WKScriptMessage, body: Body)` protocol requirement anyway.

nesium commented:
@RemiBardon Given that the flow is WKScriptMessageHandler > ViewStore > Reducer > ViewStore > View and given that the view is the entity which presents other views *and* holds a reference to the WKWebView, why do we need to pass the WKWebView through all the way?

RemiBardon commented:
Ok, I will try to present the picker from the view then.
(I had answered, but never sent the comment. As you can see in my last commit, I managed to not send the view in the actions)

RemiBardon commented:
I think I fixed everything, except

> The alternative - for this feature set - would be that we'd add an extra messageReducer that receives the JS events as actions. This could be a [forEach reducer](https://pointfreeco.github.io/swift-composable-architecture/Reducer/#reducer.foreach(state:action:environment:file:line:)). The messages should be held in an `IdentifiedArray`, also because there's a lot of index searching going on already.

and the few `#warning`s I added.

I'm not sure a `forEach` reducer would be our best option here, as some actions can make sense on an array of message IDs and not just a single one. I don't have an example in mind, it's just what the prose-core-views API offers. If you want, I can implement it.

Also, @nesium I need your opinion on this: How should I handle removing duplicates in `updateNSView`? I have two options in mind, one messy and one cleaner:
- The messy way
  1. Add `var previousState: ChatState?` to `ChatView.Coordinator`
  2. Start with `nil`, and update at the end of `if !webView.isLoading {` (inside this condition otherwise nothing would update)
  3. Replace every `do {` I wrote by `if context.coordinator.viewStore.previousState?.messages != self.viewStore.messages {` for example
- The cleaner way:
  - In createNSView, as we do
    ```swift
    webView.publisher(for: \.isLoading)
      .filter { isLoading in !isLoading }
      .prefix(1)
      .sink(receiveValue: { [viewStore] _ in
        viewStore.send(.webViewReady)
      })
      .store(in: &context.coordinator.cancellables)
    ```
    create munltiple scoped `ViewStore`s and use `viewStore.publisher.sink` to run the desired code
  - This means we have to keep strong references to the `ViewStore`s
Do you have another idea? What would you suggest?

RemiBardon commented:
After a small talk with @nesium, I'm going for the second option, but using `viewStore.publisher.my.keypath` to avoid the creation of multiple `ViewStore`s. Thanks for the tip!

RemiBardon commented:
As we changed the API in [prose-core-views v10.0.1](https://github.com/prose-im/prose-core-views/releases/tag/0.10.1), I think we can delay the `forEach` reducer to the appropriate PR.

If I didn't miss anything, I fixed everything you asked for @nesium. Is everything good now?

RemiBardon commented:
Oh, I forgot three `#warning("TODO: Throw an exception instead of `fatalError`")`. How should I handle them?

RemiBardon commented:
@valeriansaliou What do you think?
<img width="1072" alt="Screenshot 2022-08-03 at 16 16 28" src="https://user-images.githubusercontent.com/37386490/182631000-61d9bb7b-3c99-467e-9e2a-8feb46b49db8.png">
<img width="1072" alt="Screenshot 2022-08-03 at 16 16 53" src="https://user-images.githubusercontent.com/37386490/182631096-bc8b1f7c-b690-4f0f-9cec-44b225374c4c.png">

valeriansaliou commented:
LGTM, can you also add a report button which would create a pre-filled Markdown-formatted email (for the error trace) that would send this to support+macos [at] (our domain) [dot] org ?

That way it gets into Crisp and we can create a rule based on the `macos` alias tag in the email. Opening eg. Apple Mail or any default mail program, pre-filling the subject, email and body and letting the user press send is OK for me.

RemiBardon commented:
I thought about it, but I didn't want to go super fancy in this PR (and I wanted to ask you how to do it properly). Do you mind if I open an issue for this, and we get on it when we introduce logs sharing?
